transport/transporters/rabbitmq/transporter/transporter.go

```diff
@@ -0,0 +1,346 @@
+func (t *RabbitMQTransporter) StartTransporting() {
+	t.log.Info("starting transporter")
+	defer t.shutdown()
+
+	var b interface{}
+	var ok bool
+
+	for {
+		select {
+		case <-t.shutdownHandler.TerminateCtx.Done():
+			t.log.Debug("received terminateCtx cancellation")
+			return
+
+		case b, ok = <-t.inputChan:
+			// pass
+		}
+
+		select {
+		case <-t.shutdownHandler.TerminateCtx.Done():
+			t.log.Debug("received terminateCtx cancellation")
+			return
+		default:
+			// pass
+		}
+
+		if !ok {
+			t.log.Warn("input channel is closed")
+			return
+		}
+
+		genericBatch, ok := b.(*batch.GenericBatch)
+
+		if !ok {
+			panic("Batch is not a GenericBatch")
+		}
+
+		messages := genericBatch.GetPayload()
+		messagesSlice, ok := messages.([]*marshaller.MarshalledMessage)
+
+		if !ok {
+			panic("Batch payload is not a []*marshaller.MarshalledMessage")
+		}
+
+		// Begin timer
+		start := TimeSource.UnixNano()
+
+		// send to RabbitMQ with some retry logic
+		cancelled, err := t.transportWithRetry(t.shutdownHandler.TerminateCtx, messagesSlice)
+
+		// End timer and send stat
+
+		total := (TimeSource.UnixNano() - start) / int64(time.Millisecond)
+		t.statsChan <- stats.NewStatHistogram("rabbitmq_transport", "duration", total, TimeSource.UnixNano(), "ms")
+
+		if err != nil {
+			t.log.Error("max retries exceeded")
+			return
+		}
+
+		// if transportWithRetry was cancelled, then loop back around to process another potential batch or shutdown
+		if cancelled {
+			continue
+		}
+
+		t.log.Debug("successfully wrote batch")
+		t.statsChan <- stats.NewStatCount("rabbitmq_transport", "written", int64(genericBatch.NumMessages()), TimeSource.UnixNano())
+
+		// report transactions written in this batch
+		t.txnsWritten <- genericBatch.GetTransactions()
+	}
+}
+
+func (t *RabbitMQTransporter) transportWithRetry(ctx context.Context, messagesSlice []*marshaller.MarshalledMessage) (bool, error) {
+	var cancelled bool
+
+	// An operation that may fail.
+	operation := func() error {
+		select {
+		case <-ctx.Done():
+			t.log.Debug("received terminateCtx cancellation")
+			cancelled = true
+			return nil
+		default:
+		}
+
+		err := t.setupChannel(ctx)
+		if err != nil {
+			t.log.WithError(err).Error("Could not open channel to RabbitMQ")
+			return err
+		}
+
+		err = t.sendMessages(ctx, messagesSlice)
+
+		// If any errors occurred during sending then entire batch will be retried
+		if err != nil {
+			t.log.WithError(err).Error("Could not transport messages")
+			t.statsChan <- stats.NewStatCount("rabbitmq_transport", "failure", 1, TimeSource.UnixNano())
+			return err
+		}
+
+		remaining, err := t.waitForConfirmations(ctx, uint64(len(messagesSlice)))
+		// If there are no failures then all messages were sent and received
+		if err == nil {
+			t.statsChan <- stats.NewStatCount("rabbitmq_transport", "success", 1, TimeSource.UnixNano())
+			return nil
+		}
+
+		if remaining > 0 {
+			messagesSlice = messagesSlice[uint64(len(messagesSlice))-remaining:]
```

Comment:
I'm not very familiar with rabbitmq (and the library), are we guaranteed to have messages sent and acknowledged in order? Ie is it safe to assume we only need to retry `uint64(len(messagesSlice))-remaining:`?

Reply:
Rereading the example code, no.  This isn't a safe assumption and I need to redo this to use a map of deliveryTags to messages.

Reply:
Just how the golang client behaves here is a little confusing.  Hopefully the authors can explain it better for me.  https://github.com/streadway/amqp/issues/396

Reply:
It is safe to assume we will receive the acknowledgements in order, as the golang library will reorder the acknowledgements for us.  This does increase the possibility of resubmitting an already delivered message in the event of a crash, but since this is already a possibility, the user needs to be okay with that already.

With that in mind, I think the above code is correct.